src/coredump/coredump-ratelimit.c

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/statvfs.h>
+
+#include "sd-id128.h"
+
+#include "coredump-ratelimit.h"
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "missing_sched.h"
+#include "string-util.h"
+#include "time-util.h"
+#include "user-util.h"
+#include "util.h"
+
+static int exe_from_file_name(const char *file_name, char *exe) {
+        const char *p;
+        int i;
+
+        /* check if the file is a core dump file */
+        p = startswith(file_name, "core.");
+        if (!p) {
+                /* check if the file is an inprogress core dump file */
+                p = startswith(file_name, ".#core.");
+                if (!p)
+                        return -EINVAL;
+        }
+
+        /* Retrieve the process name (COMM) */
+        for (i = 0; *p != '.' && i < TASK_COMM_LEN; p++, i++){
+                exe[i] = *p;
+        }
+        exe[i] = '\0';
+        return 0;
+}
+
+static int bootid_from_file_name(const char *file_name, char *bootid) {
+        const char *p;
+        int i;
+
+        /* check if the file is a core dump file */
+        p = startswith(file_name, "core.");
+        if (!p) {
+                /* check if the file is an inprogress core dump file */
+                p = startswith(file_name, ".#core.");
+                if (!p)
+                        return -EINVAL;
+        }
+
+        /* Skip the comm field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Skip the uid field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Retrieve the Bootid */
+        for (i = 0; *p != '.' && i < SD_ID128_STRING_MAX - 1; p++, i++){
+                bootid[i] = *p;
+        }
+        bootid[i] = '\0';
+        return 0;
+}
+
+int coredump_ratelimit(const char* process_name, usec_t interval, unsigned burst) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        struct dirent *de;
+        unsigned int count = 0;
+        usec_t curr_ts;
+        sd_id128_t boot_id;
+        char sid[33];
+        int r;
+
+        /* If the process name is not given, then return -EINVAL. No rate limit happens.*/
+        if (process_name == NULL)
+                return -EINVAL;
+
+        /* either if burst or interval is zero then ratelimit is disabled */
+        if (burst == 0 || interval == 0)
+                return 0;  /* No ratelimit */
+
+        /* Whenever a daemon crashes, we compare the number of core dump files present in the
+         * /var/lib/systemd/coredump belonging to that daemon and generated within the
+         * rate limit interval with the rate limit. If the limit is reached, the core dump
+         * will not be processed further for generation.
+         */
+
+        d = opendir("/var/lib/systemd/coredump");
+        if (!d) {
+                if (errno == ENOENT)
+                        return -ENOENT;
```

Comment:
i think this should return 0, i.e. eat up the error, because ENOENT is not really one here, it just means there are no coredumps.